OptimizelySDK/Bucketing/DecisionService.cs

```diff
@@ -467,29 +490,57 @@ UserProfile userProfile
                 return;
             }
 
+            if (_userProfile == null)
+            {
+                _userProfile = userProfile;
+            }
+
             Decision decision;
-            if (userProfile.ExperimentBucketMap.ContainsKey(experiment.Id))
+            if (_userProfile.ExperimentBucketMap.ContainsKey(experiment.Id))
             {
-                decision = userProfile.ExperimentBucketMap[experiment.Id];
+                decision = _userProfile.ExperimentBucketMap[experiment.Id];
                 decision.VariationId = variation.Id;
             }
             else
             {
                 decision = new Decision(variation.Id);
             }
 
-            userProfile.ExperimentBucketMap[experiment.Id] = decision;
+            _userProfile.ExperimentBucketMap[experiment.Id] = decision;
+
+            if (!_decisionBatchInProgress)
+            {
+                SaveToUserProfileService(experiment, variation);
+            }
+        }
 
+        private void SaveToUserProfileService(Experiment experiment = null,
+            Variation variation = null
+        )
```

Comment:
What about making experiment and variation NonNull and drop ?
I see this will be called only when a new decision is made.